File: go/clock/clock.go
package clock

import (
	"fmt"
)

const TestVersion = 2

type Clock struct {
	Hours   int
	Minutes int
}

func (c Clock) String() string {
	return fmt.Sprintf("%02d:%02d", c.Hours, c.Minutes)
}

func (c Clock) Add(mins int) Clock {
	return Time(c.Hours, c.Minutes+mins)
}

func Time(h int, m int) Clock {
	h += m / 60
	m %= 60
	if m < 0 {
		h -= 1
		m += 60
	}
	h %= 24
	if h < 0 {
		h += 24
	}
	return Clock{Hours: h, Minutes: m}
}

File: haskell/clock/clock.hs
module Clock (fromHourMin, toString) where

import           Text.Printf

minsPerDay :: Integer
minsPerDay = 1440

data Time = Time Integer
  deriving Eq
instance Show Time where
  show t = printf "%02d:%02d" (hours t) (minutes t)
instance Num Time where
  (+) (Time t) (Time b) = fromInteger $ t + b
  (*) (Time t) (Time b) = fromInteger $ t * b
  abs (Time t) = Time $ abs t
  signum (Time t) = Time $ signum t
  fromInteger t = Time $ abs t `mod` minsPerDay
  negate (Time t) = fromInteger $ minsPerDay - t `mod` minsPerDay
-- 
minutes :: Time -> Integer
minutes (Time t) = t `mod` 60

hours :: Time -> Integer
hours (Time t) = t `div` 60

fromHourMin :: Integer -> Integer -> Time
fromHourMin h m = fromInteger (h * 60 + m)

toString :: Time -> String
toString = show